### PR TITLE
Native DataLoader dispatch strategy without Instrumentation

### PR DESCRIPTION
**Pull request by andimarek**
Reimplement the DataLoader dispatch strategy as a native implementation without the usage of Instrumentation.
If you are using the automatic DataLoader dispatching provided by the `DataLoaderDispatcherInstrumentation` you don't need to do anything: the behaviour is exactly the same.

# Breaking change details
This is a breaking change as it removes the `DataLoaderDispatcherInstrumentation` (and related classes) and the related method `GraphQL.doNotAddDefaultInstrumenation`.

If a user doesn't want to use the build in dispatching they can opt out with `GraphQL.doNotAutomaticallyDispatchDataLoader`.

### CODE REVIEW

File: src/main/java/graphql/GraphQL.java [hunk 1]
```diff
@@ -266,19 +267,28 @@ public Builder executionIdProvider(ExecutionIdProvider executionIdProvider) {
...
+         * As of GraphQL Java 22 there are no default instrumentations. Before that a DataLoaderDispatcherInstrumentation
+         * was added by default.
          * <p>
-         * By default a certain graphql instrumentations will be added to the mix to more easily enable certain functionality.  This
-         * allows you to stop this behavior
+         * For GraphQL Java 22 calling this method is equal to calling {@link #doNotAutomaticallyDispatchDataLoader()}
          *
          * @return this builder
          */
         public Builder doNotAddDefaultInstrumentations() {
```
Comment (bbakerman) [inline src/main/java/graphql/GraphQL.java:277]: I would deprecate this method.  Its no longer useful and semantically weird without

File: src/main/java/graphql/execution/AsyncExecutionStrategy.java [hunk 2]
```diff
@@ -63,12 +64,14 @@ public CompletableFuture<ExecutionResult> execute(ExecutionContext executionCont
             for (FieldValueInfo completeValueInfo : completeValueInfos) {
                 fieldValuesFutures.add(completeValueInfo.getFieldValueFuture());
             }
+            executionContext.getDataLoaderDispatcherStrategy().executionStrategy_onFieldValuesInfo(completeValueInfos, parameters);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/AsyncExecutionStrategy.java:67]: `executionStrategy_onFieldValuesInfo`

pretty weird naming - looks like Python

File: src/main/java/graphql/execution/AsyncExecutionStrategy.java [hunk 3]
```diff
@@ -36,6 +36,7 @@ public AsyncExecutionStrategy(DataFetcherExceptionHandler exceptionHandler) {
     @Override
     @SuppressWarnings("FutureReturnValueIgnored")
     public CompletableFuture<ExecutionResult> execute(ExecutionContext executionContext, ExecutionStrategyParameters parameters) throws NonNullableFieldWasNullException {
+        executionContext.getDataLoaderDispatcherStrategy().executionStrategy(executionContext, parameters);
         Instrumentation instrumentation = executionContext.getInstrumentation();
```
Comment (bbakerman) [inline src/main/java/graphql/execution/AsyncExecutionStrategy.java:40]: assign executionContext.getDataLoaderDispatcherStrategy() to a variable  and then use in all the places.

I feel its more readable

File: src/main/java/graphql/execution/DataLoaderDispatchStrategy.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    DataLoaderDispatchStrategy NO_OP = new DataLoaderDispatchStrategy() {
+    };
+
+
+    default void executionStrategy(ExecutionContext executionContext, ExecutionStrategyParameters parameters) {
+
+    }
+
+    default void executionStrategy_onFieldValuesInfo(List<FieldValueInfo> fieldValueInfoList, ExecutionStrategyParameters parameters) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/DataLoaderDispatchStrategy.java:20]: xxx_yyyYyyy is weird naming

Whats the point of `executionStrategy_` here - why not `onFieldValuesInfo`

File: src/main/java/graphql/execution/DataLoaderDispatchStrategy.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    default void executionStrategy_onFieldValuesException(Throwable t, ExecutionStrategyParameters parameters) {
+
+    }
+
+
+    default void executeObject(ExecutionContext executionContext, ExecutionStrategyParameters executionStrategyParameters) {
+
+    }
+
+    default void executeObject_onFieldValuesInfo(List<FieldValueInfo> fieldValueInfoList, ExecutionStrategyParameters parameters) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/DataLoaderDispatchStrategy.java:33]: ahhh different things - I see

Comment (bbakerman) [reply src/main/java/graphql/execution/DataLoaderDispatchStrategy.java:33]: still weird naming

`executeObjectOnFieldValuesInfo` is more java

File: src/main/java/graphql/execution/Execution.java [hunk 6]
```diff
@@ -76,30 +86,30 @@ public CompletableFuture<ExecutionResult> execute(Document document, GraphQLSche
...
-                .root(executionInput.getRoot())
-                .fragmentsByName(fragmentsByName)
-                .coercedVariables(coercedVariables)
-                .document(document)
-                .operationDefinition(operationDefinition)
-                .dataLoaderRegistry(executionInput.getDataLoaderRegistry())
-                .locale(executionInput.getLocale())
-                .valueUnboxer(valueUnboxer)
-                .executionInput(executionInput)
-                .build();
```
Comment (bbakerman) [inline src/main/java/graphql/execution/Execution.java:98]: mate you have to sort out your IDEA settings - why are yours so different

Comment (bbakerman) [reply src/main/java/graphql/execution/Execution.java:98]: we should check in the IDEA styles

File: src/main/java/graphql/execution/Execution.java [hunk 7]
```diff
@@ -209,6 +222,18 @@ private CompletableFuture<ExecutionResult> incrementalSupport(ExecutionContext e
...
     }
 
+    private DataLoaderDispatchStrategy createDataLoaderDispatchStrategy(ExecutionContext executionContext, ExecutionStrategy executionStrategy) {
+        if (executionContext.getDataLoaderRegistry() == EMPTY_DATALOADER_REGISTRY || doNotAutomaticallyDispatchDataLoader) {
+            return DataLoaderDispatchStrategy.NO_OP;
+        }
+        if (executionStrategy instanceof AsyncExecutionStrategy) {
+            return new PerLevelDataLoaderDispatchStrategy(executionContext);
+        } else {
+            return new FallbackDataLoaderDispatchStrategy(executionContext);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/Execution.java:232]: go over this one in the talk session - what is FallbackDataLoaderDispatchStrategy intended for ???

File: src/main/java/graphql/execution/ExecutionContext.java [hunk 8]
```diff
@@ -59,6 +60,8 @@ public class ExecutionContext {
     private final ExecutionInput executionInput;
     private final Supplier<ExecutableNormalizedOperation> queryTree;
 
+    private final AtomicReference<DataLoaderDispatchStrategy> dataLoaderDispatcherStrategy = new AtomicReference<>(DataLoaderDispatchStrategy.NO_OP);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionContext.java:63]: could be just voilatile - its a one way set no?

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 9]
```diff
@@ -136,6 +136,7 @@ public abstract class ExecutionStrategy {
     protected final DataFetcherExceptionHandler dataFetcherExceptionHandler;
     private final ResolveType resolvedType = new ResolveType();
 
+
     /**
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java:140]: yup fair enough - love me a new line ;)

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 10]
```diff
@@ -221,12 +221,14 @@ protected CompletableFuture<Map<String, Object>> executeObject(ExecutionContext
...
                 resultFutures.add(completeValueInfo.getFieldValueFuture());
             }
+            executionContext.getDataLoaderDispatcherStrategy().executeObject_onFieldValuesInfo(completeValueInfos, parameters);
             resolveObjectCtx.onFieldValuesInfo(completeValueInfos);
             resultFutures.await().whenComplete(handleResultsConsumer);
         }).exceptionally((ex) -> {
             // if there are any issues with combining/handling the field results,
             // complete the future at all costs and bubble up any thrown exception so
             // the execution does not hang.
+            executionContext.getDataLoaderDispatcherStrategy().executeObject_onFieldValuesException(ex, parameters);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java:231]: use a variable above ?  slightly faster since its a AtomicRef.get() each time

Comment (bbakerman) [reply src/main/java/graphql/execution/ExecutionStrategy.java:231]: very slight of course

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 11]
```diff
@@ -386,50 +382,51 @@ protected CompletableFuture<FetchedValue> fetchField(ExecutionContext executionC
...
+                .localContext(parameters.getLocalContext())
+                .arguments(argumentValues)
+                .fieldDefinition(fieldDef)
+                .mergedField(parameters.getField())
+                .fieldType(fieldDef.getType())
+                .executionStepInfo(executionStepInfo)
+                .parentType(parentType)
+                .selectionSet(fieldCollector)
+                .queryDirectives(queryDirectives)
+                .build();
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java:402]: this is making your PRS way longer than they really are

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 12]
```diff
@@ -386,50 +382,51 @@ protected CompletableFuture<FetchedValue> fetchField(ExecutionContext executionC
...
 
         Instrumentation instrumentation = executionContext.getInstrumentation();
 
         InstrumentationFieldFetchParameters instrumentationFieldFetchParams = new InstrumentationFieldFetchParameters(executionContext, dataFetchingEnvironment, parameters, dataFetcher instanceof TrivialDataFetcher);
         InstrumentationContext<Object> fetchCtx = nonNullCtx(instrumentation.beginFieldFetch(instrumentationFieldFetchParams,
-                executionContext.getInstrumentationState())
+            executionContext.getInstrumentationState())
         );
 
+        dataFetcher = executionContext.getDataLoaderDispatcherStrategy().modifyDataFetcher(dataFetcher);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java:413]: do we want this before or after the instrumentation one

there is no contract that the instrumentation use the original

So if we want to have a garunateed wrapper I think this should be after the one we got from instrumentation

File: src/main/java/graphql/execution/instrumentation/dataloader/PerLevelDataLoaderDispatchStrategy.java [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+    @Override
+    public void executionStrategy_onFieldValuesInfo(List<FieldValueInfo> fieldValueInfoList, ExecutionStrategyParameters parameters) {
+        int curLevel = parameters.getPath().getLevel() + 1;
+        onFieldValuesInfoDispatchIfNeeded(fieldValueInfoList, curLevel, parameters);
+    }
+
+    public void executionStrategy_onFieldValuesException(Throwable t, ExecutionStrategyParameters executionStrategyParameters) {
+        int curLevel = executionStrategyParameters.getPath().getLevel() + 1;
+        synchronized (callStack) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/PerLevelDataLoaderDispatchStrategy.java:116]: re-entrant lock as discussed elsewhere

File: src/test/groovy/example/http/HttpMain.java [hunk 14]
```diff
@@ -140,18 +134,18 @@ private void handleStarWars(HttpServletRequest httpRequest, HttpServletResponse
...
+        //         asList(new TracingInstrumentation(), dlInstrumentation)
+        // );
 
         // finally you build a runtime graphql object and execute the query
         GraphQL graphQL = GraphQL
                 .newGraphQL(schema)
                 // instrumentation is pluggable
-                .instrumentation(instrumentation)
+            // .instrumentation(instrumentation)
                 .build();
```
Comment (bbakerman) [inline src/test/groovy/example/http/HttpMain.java:149]: kill it not comment it out ??

File: src/test/groovy/graphql/GraphQLTest.groovy [hunk 15]
```diff
@@ -996,38 +994,38 @@ class GraphQLTest extends Specification {
...
+//
+//        when:
+//        def graphql = builder
+//                .build()
+//        graphql.execute('{ hello }')
+//
+//        then:
+//        queryStrategy.instrumentation instanceof DataLoaderDispatcherInstrumentation
+//    }
+//
```
Comment (bbakerman) [inline src/test/groovy/graphql/GraphQLTest.groovy:1028]: kill it dont comment it out

File: src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy [hunk 16]
```diff
@@ -65,20 +65,20 @@ class DataLoaderDispatcherInstrumentationTest extends Specification {
...
+//                .instrumentation(new SimplePerformantInstrumentation())
+//                .build()
+//        def executionInput = newExecutionInput().query('{ hero { name } }').build()
+//        when:
+//        graphQL.execute(executionInput)
+//        then:
+//        executionInput.getDataLoaderRegistry() != null
+//        def chainedInstrumentation = captureStrategy.instrumentation as ChainedInstrumentation
+////        chainedInstrumentation.instrumentations.any { instr -> instr instanceof DataLoaderDispatcherInstrumentation }
+//    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy:81]: kill dont comment out

File: src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy [hunk 17]
```diff
@@ -165,11 +165,12 @@ class DataLoaderDispatcherInstrumentationTest extends Specification {
...
         def dlRegistry = starWarsWiring.newDataLoaderRegistry()
 
-        def batchingInstrumentation = new DataLoaderDispatcherInstrumentation()
+//        def batchingInstrumentation = new DataLoaderDispatcherInstrumentation()
 
         def graphql = GraphQL.newGraphQL(starWarsWiring.schema)
                 .queryExecutionStrategy(executionStrategy)
-                .instrumentation(batchingInstrumentation).build()
+//                .instrumentation(batchingInstrumentation)
+                .build()
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy:173]: kill

File: src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderTypeMismatchTest.groovy [hunk 18]
```diff
@@ -62,7 +62,7 @@ class DataLoaderTypeMismatchTest extends Specification {
         def schema = new SchemaGenerator().makeExecutableSchema(typeDefinitionRegistry, wiring)
 
         def graphql = GraphQL.newGraphQL(schema)
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
+//                .instrumentation(new DataLoaderDispatcherInstrumentation())
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderTypeMismatchTest.groovy:65]: kill

File: src/test/groovy/graphql/execution/instrumentation/dataloader/Issue1178DataLoaderDispatchTest.groovy [hunk 19]
```diff
@@ -75,8 +75,8 @@ class Issue1178DataLoaderDispatchTest extends Specification {
 
         when:
         def graphql = TestUtil.graphQL(sdl, wiring)
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
+//                .instrumentation(new DataLoaderDispatcherInstrumentation())
+//                .instrumentation(new DataLoaderDispatcherInstrumentation())
                 .build()
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/Issue1178DataLoaderDispatchTest.groovy:80]: kill

File: src/test/groovy/graphql/execution/instrumentation/dataloader/PeopleCompaniesAndProductsDataLoaderTest.groovy [hunk 20]
```diff
@@ -184,7 +183,7 @@ class PeopleCompaniesAndProductsDataLoaderTest extends Specification {
 
         GraphQL graphQL = GraphQL
                 .newGraphQL(graphQLSchema)
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
+//                .instrumentation(new DataLoaderDispatcherInstrumentation())
                 .build()
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/PeopleCompaniesAndProductsDataLoaderTest.groovy:187]: kill

File: src/test/groovy/readme/DataLoaderBatchingExamples.java [hunk 21]
```diff
@@ -85,20 +83,20 @@ public Object get(DataFetchingEnvironment environment) {
         // available to the query and the associated DataFetchers
         //
         // In this case we use options to make it keep statistics on the batching efficiency
+        // //
+        // DataLoaderDispatcherInstrumentationOptions options = DataLoaderDispatcherInstrumentationOptions
+        //         .newOptions().includeStatistics(true);
```
Comment (bbakerman) [inline src/test/groovy/readme/DataLoaderBatchingExamples.java:88]: kill

File: src/main/java/graphql/execution/instrumentation/dataloader/PerLevelDataLoaderDispatchStrategy.java [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+import graphql.schema.DataFetcher;
+import org.dataloader.DataLoaderRegistry;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+@Internal
+public class PerLevelDataLoaderDispatchStrategy implements DataLoaderDispatchStrategy {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/PerLevelDataLoaderDispatchStrategy.java:19]: can we unit test this  eg make a series of calls to it as if it was a real engine??

File: src/main/java/graphql/GraphQL.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/GraphQL.java:277]: will remove it as discussed

File: src/main/java/graphql/execution/ExecutionStrategy.java (hunk 12 above)
Comment (andimarek) [reply src/main/java/graphql/execution/ExecutionStrategy.java:423]: good point

File: src/main/java/graphql/execution/ExecutionContext.java (hunk 8 above)
Comment (andimarek) [reply src/main/java/graphql/execution/ExecutionContext.java]: done

**Pull request merged**